tools/api_proto_breaking_change_detector/detector.py

```diff
@@ -0,0 +1,223 @@
+    def lock_file_changed(self) -> bool:
+        """Returns whether the changes between the ``before`` and ``after`` states of the proto file given in ``__init__``
+        cause a change in the detector's state file. This function is mostly used for testing, to ensure that the breaking change 
+        detector is checking all of the protobuf features we are interested in.
+
+        :return: a boolean flag indicating if the changes between ``before`` and ``after`` cause a change in the state file
+        :rtype: ``bool``
+        """
+        pass
+
+
+class ChangeDetectorError(Exception):
+    pass
+
+
+class ChangeDetectorInitializeError(ChangeDetectorError):
+    pass
+
+
+BUF_STATE_FILE = "tmp.json"
+
+
+class BufWrapper(ProtoBreakingChangeDetector):
+
+    def __init__(
+            self,
+            path_to_before: str,
+            path_to_after: str,
+            additional_args: List[str] = None) -> None:
+        """Initializes the breaking change detector, setting up any necessary config without actually running
+        the detector against any proto files.
+
+        Takes in a single protobuf as 2 files, in a ``before`` state and an ``after`` state,
+        and checks if the ``after`` state violates any breaking change rules.
+
+        :param path_to_before: absolute path to the .proto file in the before state
+        :type path_to_before: ``str``
+
+        :param path_to_after: absolute path to the .proto file in the after state
+        :type path_to_after: ``str``
+
+        :param additional_args: additional arguments for the breaking change detector CLI. May be tool-dependent.
+            Additional arguments are passed in both the "initialize" and "detect changes" steps.
+        :type additional_args: ``List[str]``
+        """
+        if not Path(path_to_before).is_file():
+            raise ValueError(f"path_to_before {path_to_before} does not exist")
+
+        if not Path(path_to_after).is_file():
+            raise ValueError(f"path_to_after {path_to_after} does not exist")
+
+        self._path_to_before = path_to_before
+        self._path_to_after = path_to_after
+        self._additional_args = additional_args
+
+    def run_detector(self) -> None:
+        """Runs the breaking change detector with the parameters given in ``__init__``. This method should populate
+        the detector's internal data such that `is_breaking` and `lock_file_changed` do not require any additional
+        invocations to the breaking change detector.
+        """
+        # 1) pull buf BSR dependencies with buf mod update (? still need to figure this out. commented out for now)
+        # 2) copy buf.yaml into temp dir
+        # 3) copy start file into temp dir
+        # 4) buf build -o tmp_file
+        # 5) copy changed file into temp dir
+        # 6) buf breaking --against tmp_file
+        # 7) check for differences (if changes are breaking, there should be none)
+
+        temp_dir = tempfile.TemporaryDirectory(
```

Comment:
Please use `with` and context managers where possible to provide reliable and easy to reason about resource reclamation.